File: fm-domain/src/main/java/io/zeitwert/fm/building/adapter/api/rest/BuildingImportExportController.java
package io.zeitwert.fm.building.adapter.api.rest;

import io.dddrive.core.ddd.model.RepositoryDirectory;
import io.zeitwert.fm.account.model.enums.CodeCurrency;
import io.zeitwert.fm.app.model.RequestContextFM;
import io.zeitwert.fm.building.adapter.api.rest.dto.BuildingTransferDto;
import io.zeitwert.fm.building.adapter.api.rest.dto.BuildingTransferElementRatingDto;
import io.zeitwert.fm.building.adapter.api.rest.dto.NoteTransferDto;
import io.zeitwert.fm.building.adapter.api.rest.dto.TransferMetaDto;
import io.zeitwert.fm.building.model.ObjBuilding;
import io.zeitwert.fm.building.model.ObjBuildingPartElementRating;
import io.zeitwert.fm.building.model.ObjBuildingPartRating;
import io.zeitwert.fm.building.model.ObjBuildingRepository;
import io.zeitwert.fm.building.model.enums.*;
import io.zeitwert.fm.collaboration.model.ObjNote;
import io.zeitwert.fm.collaboration.model.ObjNoteRepository;
import io.zeitwert.fm.collaboration.model.enums.CodeNoteType;
import io.zeitwert.fm.oe.model.ObjUserFM;
import io.zeitwert.fm.oe.model.ObjUserFMRepository;
import io.zeitwert.fm.oe.model.enums.CodeCountry;
import jakarta.servlet.ServletException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ContentDisposition;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.time.OffsetDateTime;
import java.util.List;

@RestController("buildingFileTransferController")
@RequestMapping("/rest/building/buildings")
public class BuildingImportExportController {

	private static final String AGGREGATE = "zeitwert/building";
	private static final String VERSION = "1.0";
	@Autowired
	RequestContextFM requestCtx;
	@Autowired
	private RepositoryDirectory directory;
	@Autowired
	private ObjBuildingRepository buildingRepo;
	@Autowired
	private ObjUserFMRepository userRepo;
	@Autowired
	private ObjNoteRepository noteRepo;

	@GetMapping("/{id}")
	public ResponseEntity<BuildingTransferDto> exportBuilding(@PathVariable("id") Integer id)
			throws ServletException, IOException {
		ObjBuilding building = this.buildingRepo.get(id);
		BuildingTransferDto export = this.getTransferDto(building);
		String fileName = this.getFileName(building);
		ContentDisposition contentDisposition = ContentDisposition.builder("attachment").filename(fileName).build();
		HttpHeaders headers = new HttpHeaders();
		headers.setContentDisposition(contentDisposition);
		return ResponseEntity.ok().headers(headers).body(export);
	}

	@PostMapping
	public ResponseEntity<BuildingTransferDto> importBuilding(@RequestBody BuildingTransferDto dto)
			throws ServletException, IOException {
		Integer accountId = this.requestCtx.getAccountId();
		if (accountId == null) {
			return ResponseEntity.badRequest().build();
		} else if (!AGGREGATE.equals(dto.getMeta().getAggregate())) {
			return ResponseEntity.unprocessableEntity().build();
		} else if (!VERSION.equals(dto.getMeta().getVersion())) {
			return ResponseEntity.unprocessableEntity().build();
		}
		Object userId = this.requestCtx.getUser().getId();
		OffsetDateTime timestamp = OffsetDateTime.now();
		ObjBuilding building = this.buildingRepo.create(this.requestCtx.getTenantId(), userId, timestamp);
		building.accountId = accountId;
		this.fillFromDto(building, dto);
		this.buildingRepo.store(building, userId, timestamp);
		BuildingTransferDto export = this.getTransferDto(building);
		return ResponseEntity.ok().body(export);
	}

	private BuildingTransferDto getTransferDto(ObjBuilding building) {
		TransferMetaDto meta = TransferMetaDto
				.builder()
				.aggregate(AGGREGATE)
				.version(VERSION)
				.createdByUser(building.getMeta().getCreatedByUser().getEmail())
				.createdAt(building.getMeta().getCreatedAt())
				.modifiedByUser(building.getMeta().getModifiedByUser() != null
						? building.getMeta().getModifiedByUser().getEmail()
						: null)
				.modifiedAt(building.getMeta().getModifiedAt())
				.build();
		List<BuildingTransferElementRatingDto> elements = building.currentRating.getElementList().stream().map(e -> {
			return BuildingTransferElementRatingDto
					.builder()
					.buildingPart(e.getBuildingPart().getId())
					.weight(e.getWeight())
					.condition(e.getCondition())
					.ratingYear(e.getRatingYear())
					.strain(e.getStrain())
					.strength(e.getStrength())
					.description(e.getDescription())
					.conditionDescription(e.getConditionDescription())
					.measureDescription(e.getMeasureDescription())
					.build();
		}).toList();
		List<NoteTransferDto> notes = building.notes.stream().map(note -> {
			return NoteTransferDto
					.builder()
					.subject(note.subject)
					.content(note.content)
					.isPrivate(note.isPrivate)
					.createdByUser(note.getMeta().getCreatedByUser().getEmail())
					.createdAt(note.getMeta().getCreatedAt())
					.modifiedByUser(note.getMeta().getModifiedByUser() != null
							? note.getMeta().getModifiedByUser().getEmail()
							: null)
					.modifiedAt(note.getMeta().getModifiedAt())
					.build();
		}).toList();
		BuildingTransferDto export = BuildingTransferDto
				.builder()
				.meta(meta)
				.id((Integer) building.getId())
				.name(building.name)
				.description(building.description)
				.buildingNr(building.buildingNr)
				.buildingInsuranceNr(building.insuranceNr)
				.plotNr(building.plotNr)
				.nationalBuildingId(building.nationalBuildingId)
				.historicPreservation(building.historicPreservation != null
						? building.historicPreservation.getId()
						: null)
				.street(building.street)
				.zip(building.zip)
				.city(building.city)
				.country(building.country != null ? building.country.getId() : null)
				.geoAddress(building.geoAddress)
				.geoCoordinates(building.geoCoordinates)
				.geoZoom(building.geoZoom)
				.currency(building.currency != null ? building.currency.getId() : null)
				.volume(building.volume)
				.areaGross(building.areaGross)
				.areaNet(building.areaNet)
				.nrOfFloorsAboveGround(building.nrOfFloorsAboveGround)
				.nrOfFloorsBelowGround(building.nrOfFloorsBelowGround)
				.buildingType(building.buildingType != null ? building.buildingType.getId() : null)
				.buildingSubType(building.buildingSubType != null ? building.buildingSubType.getId() : null)
				.buildingYear(building.buildingYear)
				.insuredValue(building.insuredValue)
				.insuredValueYear(building.insuredValueYear)
				.notInsuredValue(building.notInsuredValue)
				.notInsuredValueYear(building.notInsuredValueYear)
				.thirdPartyValue(building.thirdPartyValue)
				.thirdPartyValueYear(building.thirdPartyValueYear)
				.buildingPartCatalog(building.currentRating.getPartCatalog() != null
						? building.currentRating.getPartCatalog().getId()
						: null)
				.buildingMaintenanceStrategy(building.currentRating.getMaintenanceStrategy() != null
						? building.currentRating.getMaintenanceStrategy().getId()
						: null)
				.ratingStatus(building.currentRating.getRatingStatus().getId())
				.ratingDate(building.currentRating.getRatingDate())
				.ratingUser(building.currentRating.getRatingUser() != null
						? building.currentRating.getRatingUser().getEmail()
						: null)
				.elements(elements)
				.notes(notes)
				.build();
		return export;
	}

	private String getFileName(ObjBuilding building) {
		return (building.account != null ? building.account.getName() + " " : "") + building.name + ".zwbd";
	}

	private void fillFromDto(ObjBuilding building, BuildingTransferDto dto) {
		try {
			building.getMeta().disableCalc();

			ObjUserFM user = (ObjUserFM) this.requestCtx.getUser();
			OffsetDateTime now = this.requestCtx.getCurrentTime();
			building.setOwner(user);
			building.name = dto.getName();
			building.description = dto.getDescription();
			building.buildingNr = dto.getBuildingNr();
			building.insuranceNr = dto.getBuildingInsuranceNr();
			building.plotNr = dto.getPlotNr();
			building.nationalBuildingId = dto.getNationalBuildingId();
			building.historicPreservation = CodeHistoricPreservation.getHistoricPreservation(dto.getHistoricPreservation());
			building.street = dto.getStreet();
			building.zip = dto.getZip();
			building.city = dto.getCity();
			building.country = CodeCountry.getCountry(dto.getCountry());
			building.geoAddress = dto.getGeoAddress();
			building.geoCoordinates = dto.getGeoCoordinates();
			building.geoZoom = dto.getGeoZoom();
			building.currency = CodeCurrency.getCurrency(dto.getCurrency());
			building.volume = dto.getVolume();
			building.areaGross = dto.getAreaGross();
			building.areaNet = dto.getAreaNet();
			building.nrOfFloorsAboveGround = dto.getNrOfFloorsAboveGround();
			building.nrOfFloorsBelowGround = dto.getNrOfFloorsBelowGround();
			building.buildingType = CodeBuildingType.getBuildingType(dto.getBuildingType());
			building.buildingSubType = CodeBuildingSubType.getBuildingSubType(dto.getBuildingSubType());
			building.buildingYear = dto.getBuildingYear();
			building.insuredValue = dto.getInsuredValue();
			building.insuredValueYear = dto.getInsuredValueYear();
			building.notInsuredValue = dto.getNotInsuredValue();
			building.notInsuredValueYear = dto.getNotInsuredValueYear();
			building.thirdPartyValue = dto.getThirdPartyValue();
			building.thirdPartyValueYear = dto.getThirdPartyValueYear();
			final ObjBuildingPartRating rating = building.currentRating != null
					? building.currentRating
					: building.addRating(user, now);
			rating.partCatalog = CodeBuildingPartCatalog.getPartCatalog(dto.getBuildingPartCatalog());
			rating.maintenanceStrategy = CodeBuildingMaintenanceStrategy.getMaintenanceStrategy(dto.getBuildingMaintenanceStrategy());
			rating.ratingStatus = CodeBuildingRatingStatus.getRatingStatus(dto.getRatingStatus());
			rating.ratingDate = dto.getRatingDate();
			rating.ratingUser = dto.getRatingUser() != null
					? this.userRepo.getByEmail(dto.getRatingUser()).get()
					: null;
			if (dto.getElements() != null) {
				dto.getElements().forEach((dtoElement) -> {
					CodeBuildingPart buildingPart = directory.getEnumeration(CodeBuildingPart.class).getItem(dtoElement.getBuildingPart());
					ObjBuildingPartElementRating element = rating.getElement(buildingPart);
					if (element == null) {
						element = rating.addElement(buildingPart);
					}
					element.weight = dtoElement.getWeight();
					element.condition = dtoElement.getCondition();
					// element.setRatingYear(dtoElement.getRatingYear());
					element.strain = dtoElement.getStrain();
					element.strength = dtoElement.getStrength();
					element.description = dtoElement.getDescription();
					element.conditionDescription = dtoElement.getConditionDescription();
					element.measureDescription = dtoElement.getMeasureDescription();
				});
			}
			if (dto.getNotes() != null) {
				CodeNoteType noteType = CodeNoteType.getNoteType("note");
				Object noteUserId = this.requestCtx.getUser().getId();
				OffsetDateTime noteTimestamp = OffsetDateTime.now();
				dto.getNotes().forEach((dtoNote) -> {
					ObjNote note = building.addNote(noteType, noteUserId);
					note.subject = dtoNote.getSubject();
					note.content = dtoNote.getContent();
					note.isPrivate = dtoNote.getIsPrivate();
					this.noteRepo.store(note, noteUserId, noteTimestamp);
				});
			}

		} finally {
			building.getMeta().enableCalc();
			building.calcAll();
		}
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/building/adapter/api/rest/BuildingImportExportController.kt
package io.zeitwert.fm.building.adapter.api.rest

import dddrive.ddd.core.model.RepositoryDirectory
import io.zeitwert.dddrive.app.model.SessionContext
import io.zeitwert.fm.account.model.enums.CodeCurrency
import io.zeitwert.fm.building.adapter.api.rest.dto.BuildingTransferDto
import io.zeitwert.fm.building.adapter.api.rest.dto.BuildingTransferElementRatingDto
import io.zeitwert.fm.building.adapter.api.rest.dto.NoteTransferDto
import io.zeitwert.fm.building.adapter.api.rest.dto.TransferMetaDto
import io.zeitwert.fm.building.model.ObjBuilding
import io.zeitwert.fm.building.model.ObjBuildingPartElementRating
import io.zeitwert.fm.building.model.ObjBuildingPartRating
import io.zeitwert.fm.building.model.ObjBuildingRepository
import io.zeitwert.fm.building.model.enums.CodeBuildingMaintenanceStrategy
import io.zeitwert.fm.building.model.enums.CodeBuildingPart
import io.zeitwert.fm.building.model.enums.CodeBuildingPartCatalog.Enumeration.getPartCatalog
import io.zeitwert.fm.building.model.enums.CodeBuildingRatingStatus.Enumeration.getRatingStatus
import io.zeitwert.fm.building.model.enums.CodeBuildingSubType.Enumeration.getBuildingSubType
import io.zeitwert.fm.building.model.enums.CodeBuildingType.Enumeration.getBuildingType
import io.zeitwert.fm.building.model.enums.CodeHistoricPreservation.Enumeration.getHistoricPreservation
import io.zeitwert.fm.collaboration.model.ObjNoteRepository
import io.zeitwert.fm.collaboration.model.enums.CodeNoteType.Enumeration.getNoteType
import io.zeitwert.fm.oe.model.ObjUserRepository
import io.zeitwert.fm.oe.model.enums.CodeCountry.Enumeration.getCountry
import jakarta.servlet.ServletException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ContentDisposition
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.io.IOException
import java.util.function.Consumer

@RestController("buildingFileTransferController")
@RequestMapping("/rest/building/buildings")
class BuildingImportExportController {

	@Autowired
	lateinit var sessionCtx: SessionContext

	@Autowired
	lateinit var directory: RepositoryDirectory

	@Autowired
	lateinit var buildingRepo: ObjBuildingRepository

	@Autowired
	lateinit var userRepo: ObjUserRepository

	@Autowired
	lateinit var noteRepo: ObjNoteRepository

	@GetMapping("/{id}")
	@Throws(ServletException::class, IOException::class)
	fun exportBuilding(
		@PathVariable("id") id: Int,
	): ResponseEntity<BuildingTransferDto?> {
		val building = buildingRepo.get(id)
		val export = getTransferDto(building)
		val fileName = getFileName(building)
		val contentDisposition = ContentDisposition.builder("attachment").filename(fileName).build()
		val headers = HttpHeaders()
		headers.contentDisposition = contentDisposition
		return ResponseEntity.ok().headers(headers).body<BuildingTransferDto?>(export)
	}

	@PostMapping
	@Throws(ServletException::class, IOException::class)
	fun importBuilding(
		@RequestBody dto: BuildingTransferDto,
	): ResponseEntity<BuildingTransferDto?> {
		val accountId = sessionCtx.accountId
		if (accountId == null) {
			return ResponseEntity.badRequest().build<BuildingTransferDto?>()
		} else if (AGGREGATE != dto.meta?.aggregate) {
			return ResponseEntity.unprocessableEntity().build<BuildingTransferDto?>()
		} else if (VERSION != dto.meta.version) {
			return ResponseEntity.unprocessableEntity().build<BuildingTransferDto?>()
		}
		val building = buildingRepo.create()
		building.accountId = accountId
		fillFromDto(building, dto)
		buildingRepo.store(building)
		val export = getTransferDto(building)
		return ResponseEntity.ok().body<BuildingTransferDto?>(export)
	}

	private fun getTransferDto(building: ObjBuilding): BuildingTransferDto {
		val meta = TransferMetaDto(
			aggregate = AGGREGATE,
			version = VERSION,
			createdByUser = userRepo.get(building.meta.createdByUserId).email,
			createdAt = building.meta.createdAt,
			modifiedByUser = if (building.meta.modifiedByUserId != null) userRepo.get(building.meta.modifiedByUserId!!).email else null,
			modifiedAt = building.meta.modifiedAt,
		)
		val elements = building.currentRating!!
			.elementList
			.stream()
			.map { e: ObjBuildingPartElementRating ->
				BuildingTransferElementRatingDto(
					buildingPart = e.buildingPart!!.id,
					weight = e.weight,
					condition = e.condition,
					ratingYear = e.ratingYear,
					strain = e.strain,
					strength = e.strength,
					description = e.description,
					conditionDescription = e.conditionDescription,
					measureDescription = e.measureDescription,
				)
			}.toList()
		val notes = building.notes
			.stream()
			.map { noteId ->
				val note = noteRepo.load(noteId)
				NoteTransferDto(
					subject = note.subject,
					content = note.content,
					isPrivate = note.isPrivate,
					createdByUser = userRepo.get(note.meta.createdByUserId).email,
					createdAt = note.meta.createdAt,
					modifiedByUser = if (note.meta.modifiedByUserId != null) userRepo.get(note.meta.modifiedByUserId!!).email else null,
					modifiedAt = note.meta.modifiedAt,
				)
			}.toList()
		val currentRating = building.currentRating!!
		val export = BuildingTransferDto(
			meta = meta,
			id = building.id as Int,
			name = building.name,
			description = building.description,
			buildingNr = building.buildingNr,
			buildingInsuranceNr = building.insuranceNr,
			plotNr = building.plotNr,
			nationalBuildingId = building.nationalBuildingId,
			historicPreservation = building.historicPreservation?.id,
			street = building.street,
			zip = building.zip,
			city = building.city,
			country = building.country?.id,
			geoAddress = building.geoAddress,
			geoCoordinates = building.geoCoordinates,
			geoZoom = building.geoZoom,
			currency = building.currency?.id,
			volume = building.volume,
			areaGross = building.areaGross,
			areaNet = building.areaNet,
			nrOfFloorsAboveGround = building.nrOfFloorsAboveGround,
			nrOfFloorsBelowGround = building.nrOfFloorsBelowGround,
			buildingType = building.buildingType?.id,
			buildingSubType = building.buildingSubType?.id,
			buildingYear = building.buildingYear,
			insuredValue = building.insuredValue,
			insuredValueYear = building.insuredValueYear,
			notInsuredValue = building.notInsuredValue,
			notInsuredValueYear = building.notInsuredValueYear,
			thirdPartyValue = building.thirdPartyValue,
			thirdPartyValueYear = building.thirdPartyValueYear,
			buildingPartCatalog = currentRating.partCatalog?.id,
			buildingMaintenanceStrategy = currentRating.maintenanceStrategy?.id,
			ratingStatus = currentRating.ratingStatus?.id,
			ratingDate = currentRating.ratingDate,
			ratingUser = currentRating.ratingUser?.email,
			elements = elements,
			notes = notes,
		)
		return export
	}

	private fun getFileName(building: ObjBuilding): String = (if (building.account != null) building.account!!.name + " " else "") + building.name + ".zwbd"

	private fun fillFromDto(
		building: ObjBuilding,
		dto: BuildingTransferDto,
	) {
		try {
			val userId = sessionCtx.userId
			val now = sessionCtx.currentTime

			building.meta.disableCalc()

			building.ownerId = userId
			building.name = dto.name
			building.description = dto.description
			building.buildingNr = dto.buildingNr
			building.insuranceNr = dto.buildingInsuranceNr
			building.plotNr = dto.plotNr
			building.nationalBuildingId = dto.nationalBuildingId
			building.historicPreservation = getHistoricPreservation(dto.historicPreservation)
			building.street = dto.street
			building.zip = dto.zip
			building.city = dto.city
			building.country = getCountry(dto.country)
			building.geoAddress = dto.geoAddress
			building.geoCoordinates = dto.geoCoordinates
			building.geoZoom = dto.geoZoom
			building.currency = CodeCurrency.getCurrency(dto.currency)
			building.volume = dto.volume
			building.areaGross = dto.areaGross
			building.areaNet = dto.areaNet
			building.nrOfFloorsAboveGround = dto.nrOfFloorsAboveGround
			building.nrOfFloorsBelowGround = dto.nrOfFloorsBelowGround
			building.buildingType = getBuildingType(dto.buildingType)
			building.buildingSubType = getBuildingSubType(dto.buildingSubType)
			building.buildingYear = dto.buildingYear
			building.insuredValue = dto.insuredValue
			building.insuredValueYear = dto.insuredValueYear
			building.notInsuredValue = dto.notInsuredValue
			building.notInsuredValueYear = dto.notInsuredValueYear
			building.thirdPartyValue = dto.thirdPartyValue
			building.thirdPartyValueYear = dto.thirdPartyValueYear
			val rating: ObjBuildingPartRating = (
				if (building.currentRating != null) {
					building.currentRating
				} else {
					building.addRating(userId, now)
				}
			)!!
			rating.partCatalog = getPartCatalog(dto.buildingPartCatalog)
			rating.maintenanceStrategy =
				CodeBuildingMaintenanceStrategy.getMaintenanceStrategy(dto.buildingMaintenanceStrategy)
			rating.ratingStatus = getRatingStatus(dto.ratingStatus)
			rating.ratingDate = dto.ratingDate
			rating.ratingUser = if (dto.ratingUser != null) {
				userRepo.getByEmail(dto.ratingUser).get()
			} else {
				null
			}
			if (dto.elements != null) {
				dto.elements.forEach(
					Consumer { dtoElement: BuildingTransferElementRatingDto? ->
						val buildingPart = directory!!
							.getEnumeration<CodeBuildingPart>(
								CodeBuildingPart::class.java,
							).getItem(dtoElement!!.buildingPart)
						var element = rating.getElement(buildingPart)
						if (element == null) {
							element = rating.addElement(buildingPart)
						}
						element.weight = dtoElement.weight
						element.condition = dtoElement.condition
						// element.setRatingYear(dtoElement.getRatingYear());
						element.strain = dtoElement.strain
						element.strength = dtoElement.strength
						element.description = dtoElement.description
						element.conditionDescription = dtoElement.conditionDescription
						element.measureDescription = dtoElement.measureDescription
					},
				)
			}
			if (dto.notes != null) {
				val noteType = getNoteType("note")
				val noteUserId = userId
				dto.notes.forEach(
					Consumer { dtoNote: NoteTransferDto? ->
						val note = building.addNote(noteType!!, noteUserId)
						note.subject = dtoNote!!.subject
						note.content = dtoNote.content
						note.isPrivate = dtoNote.isPrivate
						noteRepo.store(note)
					},
				)
			}
		} finally {
			building.meta.enableCalc()
			building.meta.calcAll()
		}
	}

	companion object {

		private const val AGGREGATE = "zeitwert/building"
		private const val VERSION = "1.0"
	}

}
